airbyte-integrations/bigquery-destination/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java

```diff
@@ -0,0 +1,329 @@
+    return ImmutablePair.of(completedJob, null);
+  }
+
+  private static Job waitForQuery(Job queryJob) {
+    try {
+      return queryJob.waitFor();
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public StandardDiscoverSchemaOutput discover(JsonNode config) {
+    throw new RuntimeException("Not Implemented");
+  }
+
+  /**
+   * Strategy:
+   * <p>
+   * 1. Create a temporary table for each stream
+   * </p>
+   * <p>
+   * 2. Write records to each stream directly (the bigquery client handles managing when to push the
+   * records over the network)
+   * </p>
+   * <p>
+   * 4. Once all records have been written close the writers, so that any remaining records are
+   * flushed.
+   * </p>
+   * <p>
+   * 5. Copy the temp tables to the final table name (overwriting if necessary).
+   * </p>
+   *
+   * @param config - integration-specific configuration object as json. e.g. { "username": "airbyte",
+   *        "password": "super secure" }
+   * @param schema - schema of the incoming messages.
+   * @return consumer that writes singer messages to the database.
+   */
+  @Override
+  public DestinationConsumer<SingerMessage> write(JsonNode config, Schema schema) {
+
+    final BigQuery bigquery = getBigQuery(config);
+    Map<String, WriteConfig> writeConfigs = new HashMap<>();
+    final String datasetId = config.get("dataset_id").asText();
+
+    // create tmp tables if not exist
+    for (final Stream stream : schema.getStreams()) {
+      final String tableName = stream.getName();
+      final String tmpTableName = stream.getName() + "_" + Instant.now().toEpochMilli();
+
+      createTable(bigquery, datasetId, tmpTableName);
+      // https://cloud.google.com/bigquery/docs/loading-data-local#loading_data_from_a_local_data_source
+      final WriteChannelConfiguration writeChannelConfiguration = WriteChannelConfiguration
+          .newBuilder(TableId.of(datasetId, tmpTableName))
+          .setCreateDisposition(CreateDisposition.CREATE_IF_NEEDED)
+          .setSchema(com.google.cloud.bigquery.Schema.of(Field.of(COLUMN_NAME, LegacySQLTypeName.STRING)))
+          .setFormatOptions(FormatOptions.json()).build(); // new-line delimited json.
+
+      final TableDataWriteChannel writer = bigquery.writer(JobId.of(UUID.randomUUID().toString()), writeChannelConfiguration);
+
+      writeConfigs.put(stream.getName(), new WriteConfig(TableId.of(datasetId, tableName), TableId.of(datasetId, tmpTableName), writer));
+    }
+
+    // write to tmp tables
+    // if success copy delete main table if exists. rename tmp tables to real tables.
+    return new RecordConsumer(bigquery, writeConfigs, schema);
+  }
+
+  // https://cloud.google.com/bigquery/docs/tables#create-table
+  private static void createTable(BigQuery bigquery, String datasetName, String tableName) {
+    final com.google.cloud.bigquery.Schema schema = com.google.cloud.bigquery.Schema.of(Field.of(COLUMN_NAME, StandardSQLTypeName.STRING));
```

Comment:
Shouldn't we also keep track of the extraction timestamp?

Reply:
yup! adding the ids and the emitted at are on my todo checklist.